### PR TITLE
Fix/allow empty content on post and put requests

### PR DESCRIPTION
**Pull request by gpenverne**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #781
| License       | MIT

According issue with @kbsali and @teohhanhui ( #781 ) , this PR should fix error on PUT request with empty body.

### CODE REVIEW

File: src/EventListener/DeserializeListener.php [hunk 1]
```diff
@@ -66,12 +66,31 @@ public function onKernelRequest(GetResponseEvent $event)
...
+     *
+     * @param Request $request
+     *
+     * @return string
+     */
+    private function getRequestContent(Request $request) : string
+    {
+        if (in_array($request->getMethod(), [Request::METHOD_PUT, Request::METHOD_POST])
+            && $request->get('data')
+            && null === json_decode($request->getContent())
```
Comment (soyuka) [inline src/EventListener/DeserializeListener.php]: Can't we avoid `json_decode` here?

File: src/EventListener/DeserializeListener.php [hunk 2]
```diff
@@ -66,12 +66,31 @@ public function onKernelRequest(GetResponseEvent $event)
...
+     *
+     * @param Request $request
+     *
+     * @return string
+     */
+    private function getRequestContent(Request $request) : string
+    {
+        if (in_array($request->getMethod(), [Request::METHOD_PUT, Request::METHOD_POST])
+            && $request->get('data')
+            && null == $request->getContent()
```
Comment (soyuka) [inline src/EventListener/DeserializeListener.php]: maybe `empty()` is better then a non-strict equality? Not sure what `$request->getContent()` returns in the case of an empty request. From the docs it says string|Resource but if the string is `{}` it's not empty.

Comment (gpenverne) [inline src/EventListener/DeserializeListener.php]: It should work indeed. Ok for me.

File: src/EventListener/DeserializeListener.php [hunk 3]
```diff
@@ -66,12 +66,31 @@ public function onKernelRequest(GetResponseEvent $event)
         $request->attributes->set(
             'data',
             $this->serializer->deserialize(
-                $request->getContent(), $attributes['resource_class'], $format, $context
+                $this->getRequestContent($request), $attributes['resource_class'], $format, $context
```
Comment (dunglas) [inline src/EventListener/DeserializeListener.php]: Can't we just skip the deserialization if the request body is empty?

Comment (gpenverne) [inline src/EventListener/DeserializeListener.php]: If we skip this step, for example with
"        $request->attributes->set(
            'data',
            new \stdClass() (or anything you want)
We lost the loaded entity (we can't retrieve it in "data") in the case of empty put request on specific entity (for exemple PUT /dummies/1/custom_operation)

And if you just skip the $request->attributes->set, the encoding exception is thrown.

Comment (teohhanhui) [inline src/EventListener/DeserializeListener.php]: We can skip the `deserialize`. Actually I'd go as far as to skip `SerializerContextBuilder::createFromRequest` too.

> We lost the loaded entity (we can't retrieve it in "data") in the case of empty put request on specific entity (for exemple PUT /dummies/1/custom_operation)

It's already set in `ReadListener`. Don't overwrite it. :stuck_out_tongue:

File: src/EventListener/DeserializeListener.php (hunk 2 above)
Comment (teohhanhui) [inline src/EventListener/DeserializeListener.php]: `'' !== $request->getContent()`

File: src/EventListener/DeserializeListener.php [hunk 4]
```diff
@@ -66,12 +66,31 @@ public function onKernelRequest(GetResponseEvent $event)
...
     }
 
     /**
+     * Extracts the content from the Request.
+     *
+     * @param Request $request
+     *
+     * @return string
+     */
+    private function getRequestContent(Request $request) : string
```
Comment (teohhanhui) [inline src/EventListener/DeserializeListener.php]: Remove this function. It's not needed. The check is actually very simple.

**Comment by dunglas** [comment]
@gpenverne can you fix the last comments? I really want to get this merged before the release of the next RC.

**Comment by Simperfit** [comment]
Closing in favor of #805.

**Pull request closed without merging**